plenoptic/synthesize/geodesic.py

```diff
@@ -1,155 +1,340 @@
+        # storing some information
+        if self.verbose:
+            self.step_energy.append(step_energy.detach())
+            self.dev_from_line.append(
+                deviation_from_line(y.detach()))
+
+        return delta_x
+
+    def synthesize(self, max_iter=1000, learning_rate=.001, optimizer='Adam',
+                   lmbda=.1, tol=None, seed=0, verbose=True):
+        """Synthesize a geodesic via optimization.
+
+        Parameters
+        ----------
+        max_iter: int, optional
+            maximum number of steps taken by the optimization algorithm.
+        learning_rate: float, optional
+            controls the step sizes of the search algorithm.
+        optimizer: {'Adam', torch.optim.Optimizer}, optional
+            choice of algorithm that will perform the search
+            if an optimizer is passed, its `params` argument should be set to
+            `self.x`, where self refers to a previously initialized Geodesic
+            class.
+        lmbda: float, optional
+            strength of the regularizer that enforces the image range in [0, 1]
+            (strictly positive by default)
+        tol: float, optional
+            tolerance threshold used to terminate algorithm before `max_iter`
+            if the optimization stopped making progress
+        seed: int, optional
+            set the random number generator
+        verbose: bool, optional
+            storing information along the run of the optimization algorithm
+        """
+        self.lmbda = lmbda
+        self.verbose = verbose
+        if tol is None:
+            tol = self.pixelfade.norm() / 1e4 * (1 + 5 ** .5) / 2
+        print(f"\n threshold for delta_x, tolerance = {tol:.5e}")
 
         torch.manual_seed(seed)
-        self.optimizer = optim.Adam([self.x], lr=learning_rate, amsgrad=True)
-
-        self.loss = []
-        self.t1 = time.time()
-
-        for i in range(max_iter):
-            loss = self._optimizer_step(i, max_iter)
-            self.loss.append(loss.item())
-
-        self.gamma.copy_(self.img.data.squeeze()).add_(0.5).mul_(255)
-
-        return self.gamma
+        if optimizer == 'Adam':
+            self.optimizer = optim.Adam([self.x],
+                                        lr=learning_rate, amsgrad=True)
+        elif isinstance(optimizer, optim.Optimizer):
+            self.optimizer = optimizer
+
+        delta_x = torch.ones(1)
+        i = 0
+        with tqdm(range(max_iter)) as pbar:
+            # project onto set of representational geodesics
+            while delta_x > tol and i < max_iter:
+                delta_x = self._optimizer_step(pbar)
+                i += 1
+        self.populate_geodesic()
+
+    def plot_loss(self):
+        fig, ax = plt.subplots(1, 1)
+        ax.plot(self.loss)
+        ax.set(yscale='log',
+               xlabel='iter step',
+               ylabel='loss value')
+        return fig
+
+    def populate_geodesic(self):
```

Comment:
should this be a public method? will a user ever call it?

Reply:
agreed, it can be made private